pkg/resource/rule/hooks.go

```diff
@@ -0,0 +1,144 @@
+	svcapitypes "github.com/aws-controllers-k8s/eventbridge-controller/apis/v1alpha1"
+)
+
+type validationError struct {
+	field   string
+	message string
+}
+
+func (v validationError) Error() string {
+	return fmt.Sprintf("invalid Spec: %q: %s", v.field, v.message)
+}
+
+func newValidationError(field, message string) validationError {
+	return validationError{
+		field:   field,
+		message: message,
+	}
+}
+
+func validateRuleSpec(spec v1alpha1.RuleSpec) error {
+	var match bool
+	if s := spec.State; s != nil {
+		allowedValues := svcsdk.RuleState_Values()
+		for _, v := range allowedValues {
+			if *s == v {
+				match = true
+			}
+		}
+		if !match {
+			return newValidationError(
+				"spec.state",
+				fmt.Sprintf("supported states: %v", allowedValues),
+			)
+		}
+	}
+
+	emptyPattern := func() bool {
+		return spec.EventPattern == nil || *spec.EventPattern == ""
+	}
+
+	emptySchedule := func() bool {
+		return spec.ScheduleExpression == nil || *spec.ScheduleExpression == ""
+	}
+
+	if emptySchedule() && emptyPattern() {
+		return newValidationError(
+			"spec",
+			fmt.Sprintf("at least one of %q or %q must be specified",
+				"spec.eventPattern", "spec.scheduleExpression"),
+		)
+	}
+
+	// until code-gen can generate required markers for custom_field
+	for _, t := range spec.Targets {
+		arn := t.ARN
+		id := t.ID
+
+		if arn == nil || *arn == "" || id == nil || *id == "" {
+			return newValidationError(
+				"spec.targets",
+				fmt.Sprintf("%q and %q must be specified for each target", "arn", "id"),
+			)
+		}
+	}
+
+	return nil
+}
+
+// setResourceAdditionalFields will set the fields that are not returned by
+// DescribeRule calls
+func (rm *resourceManager) setResourceAdditionalFields(
+	ctx context.Context,
+	ko *svcapitypes.Rule,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.setResourceAdditionalFields")
+	defer func() { exit(err) }()
+
+	if ko.Status.ACKResourceMetadata != nil && ko.Status.ACKResourceMetadata.ARN != nil &&
+		*ko.Status.ACKResourceMetadata.ARN != "" {
+		// Set event data store tags
+		ko.Spec.Tags, err = rm.getTags(ctx, string(*ko.Status.ACKResourceMetadata.ARN))
+		if err != nil {
+			return err
+		}
+
+		ko.Spec.Targets, err = rm.getTargets(ctx, *ko.Spec.Name, *ko.Spec.EventBusName)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func equalStrings(a, b *string) bool {
+	if a == nil {
+		return b == nil || *b == ""
+	}
+
+	if a != nil && b == nil {
+		return false
+	}
+
+	return (*a == "" && b == nil) || *a == *b
+}
+
+func equalZeroString(a *string) bool {
+	return equalStrings(a, aws.String(""))
+}
+
+func customPreCompare(
+	delta *ackcompare.Delta,
+	desired *resource,
+	latest *resource,
+) {
+	if len(desired.ko.Spec.Tags) != len(latest.ko.Spec.Tags) {
+		delta.Add("Spec.Tags", desired.ko.Spec.Tags, latest.ko.Spec.Tags)
+	}
+
+	if !equalTags(desired.ko.Spec.Tags, latest.ko.Spec.Tags) {
+		delta.Add("Spec.Tags", desired.ko.Spec.Tags, latest.ko.Spec.Tags)
+	}
```

Comment:
I just opened https://github.com/aws-controllers-k8s/community/issues/1658 to solve this in a general way, so that each controller doesn't require their own implementation like this.

Reply:
Thx, again non-blocking I assume.